include/experimental/__p1673_bits/blas1_vector_sum_of_squares.hpp

```diff
@@ -0,0 +1,114 @@
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software. //
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef LINALG_INCLUDE_EXPERIMENTAL___P1673_BITS_BLAS1_VECTOR_SUM_OF_SQUARES_HPP_
+#define LINALG_INCLUDE_EXPERIMENTAL___P1673_BITS_BLAS1_VECTOR_SUM_OF_SQUARES_HPP_
+
+#include <cmath>
+#include <cstdlib>
+
+namespace std {
+namespace experimental {
+inline namespace __p1673_version_0 {
+namespace linalg {
+
+// Scaled sum of squares of a vector's elements
+template<class Scalar>
+struct sum_of_squares_result {
+  Scalar scaling_factor;
+  Scalar scaled_sum_of_squares;
+};
+
+template<class in_vector_t,
+         class Scalar>
+sum_of_squares_result<Scalar> vector_sum_of_squares(
+    in_vector_t x,
+    sum_of_squares_result<Scalar> init)
+{
+  using std::abs;
+
+  if (x.extent(0) == 0) {
+    return init;
+  }
+
+  // Rescaling, as in the Reference BLAS DNRM2 implementation, avoids
+  // unwarranted overflow or underflow.
+
+  Scalar scale = init.scaling_factor;
+  Scalar ssq = init.scaled_sum_of_squares;
+  for (ptrdiff_t i = 0; i < x.extent(0); ++i) {
+    if (abs(x(i)) != 0.0) {
+      const auto absxi = abs(x(i));
+      const auto quotient = scale / absxi;
+      if (scale < absxi) {
+          ssq = Scalar(1.0) + ssq * quotient * quotient;
```

Comment:
Again, we might need a separate implementation of `vector_sum_of_squares` for non-floating-point types, since we might not always be able to convert `1.0` to `Scalar`.